Public/Src/App/Bxl/AzureDevOpsListener.cs

```diff
@@ -0,0 +1,212 @@
+                            m_console.WriteOutputLine(MessageLevel.Info, $"##vso[task.setprogress value={currentProgress};]Pip Execution phase");
+                        }
+
+                        break;
+                    }
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void OnAlways(EventWrittenEventArgs eventData)
+        {
+        }
+
+        /// <inheritdoc />
+        protected override void OnVerbose(EventWrittenEventArgs eventData)
+        {
+            switch (eventData.EventId)
+            {
+                case (int)EventId.CacheMissAnalysis:
+                    {
+                        var payload = eventData.Payload;
+
+                        m_buildViewModel.BuildSummary.CacheSummary.Entries.Add(
+                            new CacheMissSummaryEntry
+                            {
+                                PipDescription = (string)payload[0],
+                                Reason = (string)payload[1],
+                                FromCacheLookup = (bool)payload[2],
+
+                            }
+                        );
+                    }
+                    break;
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void OnCritical(EventWrittenEventArgs eventData)
+        {
+            LogAzureDevOpsIssue(eventData, "error");
+        }
+
+        /// <inheritdoc />
+        protected override void OnError(EventWrittenEventArgs eventData)
+        {
+            LogAzureDevOpsIssue(eventData, "error");
+
+            switch (eventData.EventId)
+            {
+                case (int)EventId.PipProcessError:
+                    {
+                        var payload = eventData.Payload;
+
+                        m_buildViewModel.BuildSummary.PipErrors.Add(new BuildSummaryPipDiagnostic
+                        {
+                            SemiStablePipId = $"Pip{((long)eventData.Payload[0]):X16}",
+                            PipDescription = (string)eventData.Payload[1],
+                            SpecPath = (string)eventData.Payload[2],
+                            ToolName = (string)eventData.Payload[4],
+                            ExitCode = (int)eventData.Payload[7],
+                            Output = (string)eventData.Payload[5],
+                        }); 
+                    }
+                    break;
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void OnWarning(EventWrittenEventArgs eventData)
+        {
+            LogAzureDevOpsIssue(eventData, "warning");
+        }
+
+        /// <inheritdoc />
+        protected override void Output(EventLevel level, int id, string eventName, EventKeywords eventKeywords, string text, bool doNotTranslatePaths = false)
+        {
+        }
+
+        private void LogAzureDevOpsIssue(EventWrittenEventArgs eventData, string eventType)
+        {
+            var builder = new StringBuilder();

```

Comment:
>StringBuilder(); [](start = 30, length = 16)

should we use the pooled StringBuilder here?

Reply:
I thought it was okay since this is only for errors.

---
In reply to: [316889969](https://github.com/microsoft/BuildXL/pull/784#discussion_r316889969) [](ancestors = 316889969)